### PR TITLE
[Feature] Support function previous_day

### LINKED ISSUE #19553: [function]next_day/previous_day need to be supported.
## Feature request

it is necessary for user to check next mon/tue/wed... day. here is my proposal:

### description

Returns the date of the first specified DOW (day of week) that occurs after the input date.

### syntax

```
NEXT_DAY( <date_or_datetime_expr> , <dow_string> )
```
in which, 
`date_or_datetime_expr` can be date or datetime type
`dow_string` specifies the day of week.
Validvalues include a number of abbreviations:
Sunday, Su, Sun
Monday, Mo, Mon
Tuesday, Tu, Tue
Wednesday, We, Wed
Thursday, Th, Thu
Friday,  Fr, Fri
Saturday,  Sa, Sat

and return a date.

### PR DESCRIPTION
**Pull request by mlimwxxnn**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #19553

## Problem Summary(Required) ：

NEW FUNCTION: `previous_day`

Test sql:
```sql
CREATE TABLE IF NOT EXISTS `previous_day_test` (
  `d1` DATE,
  `d2` DATETIME
)
DISTRIBUTED BY HASH(`d1`)
PROPERTIES(
  "replication_num" = "1"
);

INSERT INTO `previous_day_test`(d1, d2)
VALUES ('2023-02-28', '2023-02-28 20:13:14'),
       ('2023-03-01', '2023-03-01 20:13:14'),
       ('2024-02-29', '2024-02-29 20:13:14'),
       ('2024-03-01', '2024-03-01 20:13:14');

CREATE TABLE IF NOT EXISTS `dow_test` (
  `d3` DATE,
  `dow_2` CHAR(20),
  `dow_3` VARCHAR(30),
  `dow_full` STRING
)
DISTRIBUTED BY HASH(`d3`)
PROPERTIES(
  "replication_num" = "1"
);

INSERT INTO `dow_test`(d3, dow_2, dow_3, dow_full)
VALUES ('2023-04-01', 'Mo', 'Mon', 'Monday'),
       ('2023-04-02', 'Tu', 'Tue', 'Tuesday'),
       ('2023-04-03', 'We', 'Wed', 'Wednesday'),
       ('2023-02-28', 'Th', 'Thu', 'Thursday'),
       ('2024-02-29', 'Fr', 'Fri', 'Friday');
```
Test results:
```Plain Text
MySQL > select previous_day('2023-04-09', 'Sunday');
+--------------------------------------+
| previous_day('2023-04-09', 'Sunday') |
+--------------------------------------+
| 2023-04-02                           |
+--------------------------------------+

MySQL > select previous_day('2023-04-09', 'Wed');
+-----------------------------------+
| previous_day('2023-04-09', 'Wed') |
+-----------------------------------+
| 2023-04-05                        |
+-----------------------------------+

MySQL > select previous_day('2023-04-09 20:08:11', 'Fr');
+-------------------------------------------+
| previous_day('2023-04-09 20:08:11', 'Fr') |
+-------------------------------------------+
| 2023-04-07                                |
+-------------------------------------------+

MySQL > select d1, previous_day(d1, 'Tuesday'), d2, previous_day(d2, 'Tuesday') from previous_day_test order by d1;
+------------+-----------------------------+---------------------+-----------------------------+
| d1         | previous_day(d1, 'Tuesday') | d2                  | previous_day(d2, 'Tuesday') |
+------------+-----------------------------+---------------------+-----------------------------+
| 2023-02-28 | 2023-02-21                  | 2023-02-28 20:13:14 | 2023-02-21                  |
| 2023-03-01 | 2023-02-28                  | 2023-03-01 20:13:14 | 2023-02-28                  |
| 2024-02-29 | 2024-02-27                  | 2024-02-29 20:13:14 | 2024-02-27                  |
| 2024-03-01 | 2024-02-27                  | 2024-03-01 20:13:14 | 2024-02-27                  |
+------------+-----------------------------+---------------------+-----------------------------+

MySQL > select d3, dow_2, previous_day(d3, dow_2), dow_3, previous_day(d3, dow_3) from dow_t
est order by d3;
+------------+-------+-------------------------+-------+-------------------------+
| d3         | dow_2 | previous_day(d3, dow_2) | dow_3 | previous_day(d3, dow_3) |
+------------+-------+-------------------------+-------+-------------------------+
| 2023-02-28 | Th    | 2023-02-23              | Thu   | 2023-02-23              |
| 2023-04-01 | Mo    | 2023-03-27              | Mon   | 2023-03-27              |
| 2023-04-02 | Tu    | 2023-03-28              | Tue   | 2023-03-28              |
| 2023-04-03 | We    | 2023-03-29              | Wed   | 2023-03-29              |
| 2024-02-29 | Fr    | 2024-02-23              | Fri   | 2024-02-23              |
+------------+-------+-------------------------+-------+-------------------------+

MySQL > select d1, dow_full, previous_day(d1, dow_full) from previous_day_test a join dow_test b on a.d1 = b.d3;
+------------+----------+----------------------------+
| d1         | dow_full | previous_day(d1, dow_full) |
+------------+----------+----------------------------+
| 2024-02-29 | Friday   | 2024-02-23                 |
| 2023-02-28 | Thursday | 2023-02-23                 |
+------------+----------+----------------------------+
```

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [x] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

File: be/src/exprs/time_functions.cpp [hunk 1]
```diff
@@ -2106,4 +2106,41 @@ std::string TimeFunctions::info_reported_by_time_slice = "time used with time_sl
...
+DEFINE_BINARY_FUNCTION_WITH_IMPL(previous_dayImpl, datetime, dow) {
+    int datetime_weekday = ((DateValue)datetime).weekday();
+    int dow_weekday = weekday_from_dow(dow.to_string());
+    if (dow_weekday != -1) {
+        return (DateValue)timestamp_add<TimeUnit::DAY>(datetime, -((6 + datetime_weekday - dow_weekday) % 7 + 1));
+    } else {
+        throw std::runtime_error(dow.to_string() + " not supported in previous_day dow_string");
+    }
+}
+DEFINE_TIME_CALC_FN(previous_day, TYPE_DATETIME, TYPE_VARCHAR, TYPE_DATE);
```
Comment (silverbullet233) [inline be/src/exprs/time_functions.cpp:2144]: same as https://github.com/StarRocks/starrocks/pull/21135#discussion_r1161455665

Comment (mlimwxxnn) [reply be/src/exprs/time_functions.cpp]: done.

File: gensrc/script/functions.py [hunk 2]
```diff
@@ -422,6 +422,7 @@
     [50371, 'date_slice', 'DATE', ['DATE', 'INT', 'VARCHAR', 'VARCHAR'], 'TimeFunctions::time_slice', 'TimeFunctions::time_slice_prepare', 'TimeFunctions::time_slice_close'],
     [50372, 'time_slice', 'DATETIME', ['DATETIME', 'INT', 'VARCHAR'], 'TimeFunctions::time_slice', 'TimeFunctions::time_slice_prepare', 'TimeFunctions::time_slice_close'],
     [50373, 'time_slice', 'DATETIME', ['DATETIME', 'INT', 'VARCHAR', 'VARCHAR'], 'TimeFunctions::time_slice', 'TimeFunctions::time_slice_prepare', 'TimeFunctions::time_slice_close'],
+    [50401, 'previous_day', 'DATE', ['DATETIME', 'VARCHAR'], 'TimeFunctions::previous_day'],
```
[the remark this answers is not in the document]
Comment (mlimwxxnn) [reply gensrc/script/functions.py]: I have fixed the problems, including next_day function.

**Comment by mlimwxxnn** [comment]
@imay Hi, could you help with review this?

**Review by imay — APPROVED** [review]

**Review by silverbullet233 — APPROVED** [review]

**Pull request merged**